### PR TITLE
Feature/watch your step

### PR DESCRIPTION
**Pull request by kieranricardo**
This pull request contains an implementation of the Watch Your Step Algorithm (https://arxiv.org/pdf/1710.09599.pdf).

The current implementation is functional but it lacks unit tests and only has a bare-bones demo.

Reviewer checklist:

- [ ] The demo works correctly and replicates the classification results from the paper
- [ ] The `AdjacencyPowerGenerator` correctly creates a tensorflow dataset of rows of the the adjacency powers
- [ ] The `graph_log_likelihood` implements the graph log likelihood loss function correctly
- [ ] The tensorflow function in `tf_utils.py` are implemented correctly
- [ ] The `WatchYourStep` and `AttentiveWalk` correctly implement the trainable embeddings and attention functionality described in the paper

### CODE REVIEW

File: demos/embeddings/watch-your-step-cora-demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
...7v29PIShOBR3u4lNZwGXlKvfAJ6ktGY3s+QLK+bBLLfPOJGDFkWqLXCGQGfF+AzBEC7kOV/SICaQpIg+ZaeAeGYe8FGNduwbPSyFkB8CQy91OsC/TcRzgg1OLjF20qSE05/nhWPfgg0hmkQU4IJh5xRM+fDSjDzcG+Dx10AFDHRm7nNk7hNPZg9qDvVR7NzXEOPfTPrF7dCgiiUYtjj53C7373ecLh3dyV/R/GzgSpR/O/SCl/KIT49m4czwg+oZjMFMYxnlZaMDAppXQrfbZ/N6I1NZRMnkz7+vWESkvRNI1gUTFe1qbmgp/QYRSgofadxgTr0DWHtKNYAUm7kPe6Z5Dx/MwpWEq5vxn0JlZky9jL0vFpinKvD/gKteZqZwMXtALlxrtfUJXSAkLQgQFkcDQfbsAkHEyTsU2mTGpjwqUTMSas5Mzjv0L9ltP4TvQsxlkNvO5OZ1N6LA9tOpk5RUuZEl6N51rUxuYz25zGwQF4JeFCJoXt8ytXYiHA85Rp4ID3xIdiJmZRLDZy16NIhyfGKRPCFRmY7VcEgh7Eb4b4j0EmKBLwG3MeLTX3EKec0aaBMX1vQE3qlzTBa0klyVSblXibfZjdKTxNI5RO0BYqJWX4ifkjCOmxpaCKyc2ryaRG4fo1TOHiSJErk6ldQieh424xkXFNEV18EnQXb5qPrBQ0OeAzIVgBCQe8LwHPCFgmIQYkJKPm17J5Zhi3RcdL2ki/hgwqkgXL0/jGRXoCVB46SjgYYI9Fi1h84400DiJmHaqsZP6FF/b8HSRIjO6ev7du1+i9/TEeYSazhlzgZTIO06b9lo6OTM+z2ttT3H33e4wfX8AVVxwy6PM+qdjmMlcIcacQoufjKaXsZ98upbzxoxjYCD55MDGpopoyyv7jAQpUifKwX/yCYEkJ8eZm4k1NJFtbmXX66TTtMQ9D9CowGMJBy5Vusq7FWx374HgmQT1BY6oaT+p02mF8ehvNLqRIESZMKf27io/OKbP3tW2XOQHUo0Oq9+baCig2BBoWcelHAoZm0zlqPN+a+jfmXRIiPNEiRIzSsM6YPebg99uEjSxTjM0EyJDNRljcsh//rD+dhxuP45XO6fygtZMtTgYjGcfTDES+FCVVKqR5Hn5kj1V9oaZM/1JSzfNVpippBTTYL9frNdlShn79AlT6aYhfCjINWghkALJvUtZ1HO
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/embeddings/watch-your-step-cora-demo.ipynb]: - `training_labels` is used in calculating the `predictions` (unless that's what you're refferring to - hard to tell with Jupyter nbs)
- the predictions section replicates (or at least is supposed to!) Section 4.3 of the paper. The test labels are zeroed out and the Watch Your Step embeddings are trained in a completely unsupervised fashion, so no information about the test labels is leaked into the prediction step so this should still be valid :)

[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/embeddings/watch-your-step-cora-demo.ipynb]: nice catch

File: stellargraph/layer/watch_your_step.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            model.get_layer("WATCH_YOUR_STEP_RIGHT_EMBEDDINGS")
+            .kernel.numpy()
+            .transpose(),
+        ]
+    )
+
+    return embeddings
+
+
+@experimental(reason="lack of unit tests")
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/watch_your_step.py]: good point, I'll make the issue

File: demos/embeddings/watch-your-step-cora-demo.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/embeddings/watch-your-step-cora-demo.ipynb]: ah my description is confusing to say the least! yep you're right, I'll edit the notebook for clarity

[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/embeddings/watch-your-step-cora-demo.ipynb]: yep that's what I've implemented. Agreed the description is confusing, reading the description does sound like your first matrix above but plugging that into their equation would result in junk! 

the one-hot-encoded label approach is the only way I could inteprate their equation/description that actually made sense. Interestingly, if you train the embeddings for longer the classification accuracy decreases and you obtain the accuracy mentioned in the paper
